Blog post about first steps on SCM integration (#294)
openSUSE/obs-landing: _posts/development/2021-05-04-scm-integration.md

---
layout: post
title: Heading for a GitHub and GitLab Workflow Integration in OBS
category: development
---

Most of the code implemented nowadays is managed collaboratively
through source code management (SCM) systems like GitHub, GitLab or Pagure.
An increasing number of people want to handle their package sources with those systems too
and while there is an integration between SCMs and OBS, it could definitely be improved.
In today's post, we are sharing how we began to move to meet that need.

## SCM to Collaborate, OBS to Build

We have started to integrate GitHub and GitLab for a scenario that seems to be the 
most expected one among OBS users: managing the package sources in an SCM system
and use OBS to build the source code and report back the results to GitHub or GitLab.

> We propose the following steps to achieve this goal. Give them a try!

Reviewer: I think you should talk a little bit more about which process you want to make possible with those changes.

- every PR/MR is triggering a package branch on OBS
- OBS is reporting back the build status of the branched package to the PR/MR

Then dive into the steps that people can do to try this. I know you have described this below but it's way better to tell people up-front what this is about.

Author: I added two sentences paragraph the whole idea: https://github.com/openSUSE/obs-landing/pull/294/commits/7bf243757183aee3beaac07b62a3dc9a1a169174#diff-643f2999b778582f99173bf8c79448e41f5e21eb147322f94705781cf890c564R19-R21

### 1. Set a Configuration File

In the root directory of your GitHub/GitLab repository, create a directory `.obs` which contains the configuration file `workflows.yml`.

The content of the file should look like this:

```
workflow1:
  steps:
    - branch_package:
        source_project: home:Iggy
        source_package: ctris
        destination_project_prefix: home:Iggy:PR- 
        destination_package: ctris
```

For now, OBS only accepts one workflow with a single step: `branch_package`. Adapt the names of the source and destination projects and packages to your needs.

The step `branch_package` indicates to OBS that you want to branch the project/package given by `source_project` and `source_package` fields.

The resulting branched project/package will be named after the fields `destination_project_prefix` and `destination_package`.
For example, the branched project name could look like `home:Iggy:PR-1234` where `1234` is added by OBS after the prefix and corresponds to the pull/merge request number.

### 2. Create Project and Package in OBS

Apart from the repository inside an SCM system, you have to create an analogous project/package inside OBS.

> For example, for a repository called `ctris` in GitHub/GitLab you might have a `home:Iggy/ctris` project and package in OBS.

Reviewer: This makes it sound as if those names have anything to do with this. But you can have a package called `hans` on OBS and a repository called `franz` on your SCM. There is no relation by name.

Way more important than the name is the whole integration with [obs-service-tar_scm](https://github.com/openSUSE/obs-service-tar_scm) and it's ability to fetch the sources from the PR/MR source fork/branch.

Author: Now I use different names.

The explanation about the integration with `obs-service-tar_scm` is still pending.

### 3. Create a Workflow Token in OBS

When some events happen in your GitHub/GitLab repository, you want to trigger some actions in OBS.
To do so, you need to generate a token in OBS and GitHub/GitLab will use it to trigger the actions on your behalf.

The token represents your user account and the kind of action/operation you want to perform, which is `workflow` in this case.

You can create a token in OBS using `osc`:

```
osc token --create --operation workflow scm_token=ssx_j2du52Ozs8oZAtxq3K0bdy9TEp8iUy4JilTs

> ```

Reviewer: Please warn people about the consequences of leaking the OBS token secrect.

Author: Now there is a WARNING paragraph :)
https://github.com/openSUSE/obs-landing/pull/294/commits/7bf243757183aee3beaac07b62a3dc9a1a169174#diff-643f2999b778582f99173bf8c79448e41f5e21eb147322f94705781cf890c564R76-R77

Adding the `scm_token` part is not really needed to generate the token. It is just an extra value containing the GitHub/GitLab token
that OBS have to use to report back to the SCM system.

> 

Reviewer: You are missing the step before this one to create personal access tokens on the SCM.

Author: It was explained later. But I moved the explanation to the second step, so it's clear from the beginning.

- the result of the building process will be reported back to GitHub/GitLab. 

### 6. OBS Reports Back the Result to the SCM System

OBS will report the result of the workflow steps back to GitHub/GitLab.
To do so, OBS needs to know the `scm_token` that represents your user account on the SCM system's side, as explained in step 3.

Check [GitHub's](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token) and [GitLab's](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html#creating-a-personal-access-token) documentation to learn how to create a personal access token.

As soon as all the steps in the workflow run, in the checks section of your Pull Request or Merge Request, you could see a new check with three possible states:
- pending (yellow)
- success (green)
- failed or failure (red)

<figure>
  <img src="/images/posts/sprint_95_checks_github.png" alt="Checks GitHub" />
  <figcaption>Checks GitHub</figcaption>
</figure>

<figure>
  <img src="/images/posts/sprint_95_checks_gitlab.png" alt="Checks GitLab" />
  <figcaption>Checks GitLab</figcaption>
</figure>

> 

Reviewer: You should also add a paragraph about all the limitations of this implementation that we already know.

- we don't handle PR syncs like new commits/force push etc. yet
- we only report back build results for one repository/arch (right?)
- what happens if I manually click rebuild on OBS?
- more?

And additionally maybe you want to give an outlook on what's more to come?

- token UI
- branch/event filtering
- more types of steps
- reporting for whole projects with many repositories
- maybe even mention outgoing webhooks

Just to get people excited for this journey. This is the first time they hear about this.

---
layout: post

> title: Heading for an SCMs Integration in OBS

Reviewer 2: ```suggestion
title: Heading for a Better SCM Integration in OBS
```

category: development
---

Most of the code implemented nowadays is managed collaboratively
through source code management (SCM) systems like GitHub, GitLab or Pagure.

> More and more people want to handle their package sources with those systems too,

Reviewer 2: ```suggestion
An increasing number of people want to handle their package sources with those systems too,
```

> what leads us to detect a lack of better integration between SCMs and OBS.

Reviewer 2: ```suggestion
and while there is an integration between SCMs and OBS, it could be definitely improved.
```

In today's post, we are sharing how we began to move to meet that need.

## SCM to Collaborate, OBS to Build

We have started to integrate GitHub and GitLab for a scenario that seems to be the 
most expected one among OBS users: managing the package sources in an SCM system
and use OBS to build the source code and report back the results to GitHub or GitLab.

We propose the following steps to achieve this goal. Give them a try!

### 1. Set a Configuration File

> Add the configuration file `.obs/workflows.yml` in the root directory of your GitHub or GitLab repository.

Reviewer 2: I would be extra clear here about the file, so even telling people to create the file in the `.obs` folder.

```suggestion
In the root directory of your GitHub/GitLab repository, create a directory `.obs` which contains the configuration file `workflows.yml`.
```

### 2. Create Project and Package in OBS

Apart from the repository inside an SCM system, you have to create an analogous project/package inside OBS.

For example, for a repository called `ctris` in GitHub/GitLab you might have a `home:Iggy/ctris` project and package in OBS.

### 3. Create a Workflow Token in OBS

When some events happen in your GitHub/GitLab repository, you want to trigger some actions in OBS.
To do so, you need to generate a token in OBS and GitHub/GitLab will use it to trigger the actions on your behalf.

The token represents your user account and the kind of action/operation you want to perform, which is `workflow` in this case.

You can create a token in OBS using `osc`:

```
osc token --create --operation workflow scm_token=ssx_j2du52Ozs8oZAtxq3K0bdy9TEp8iUy4JilTs
```

Adding the `scm_token` part is not really needed to generate the token. It is just an extra value containing the GitHub/GitLab token
that OBS have to use to report back to the SCM system.

### 4. Set a Webhook

By setting a webhook in GitHub or GitLab, you can specify the endpoint you want the SCM system to reach when certain events happen.

#### Set a GitHub Webhook

> In GitHub you can go to *Settings -> Webhooks*.

Reviewer 2: ```suggestion
In GitHub, you can go to *Settings -> Webhooks*.
```

<figure>
  <img src="/images/posts/sprint_95_github_webhook_form.png" alt="GitHub Webhook Form" />
  <figcaption>GitHub Webhook Form</figcaption>
</figure>

You have to fill in the form with
- *Payload URL*: the OBS endpoint to trigger the workflow operation (https://build.opensuse.org/trigger/workflow?id=0). Ensure you replace the `id` with the one obtained in the token creation (step 3).
- *Content type*: please choose `application/json`.
- *Secret*: the OBS's token string obtained in step 3.
- *Enable SSL verification*.
- *Let me select individual events*. We just need you to select `Pull requests`.

As soon as you add the webhook, OBS will start receiving requests for every Pull Request happening in GitHub.

#### Set a GitLab Webhook

> In GitLab you can go to *Settings -> Webhooks*.

Reviewer 2: ```suggestion
In GitLab, you can go to *Settings -> Webhooks*.
```

<figure>
  <img src="/images/posts/sprint_95_gitlab_webhook_form.png" alt="GitLab Webhook Form" />
  <figcaption>GitLab Webhook Form</figcaption>
</figure>

You have to fill in the form with
- *URL*: the OBS endpoint to trigger the workflow operation (https://build.opensuse.org/trigger/workflow?id=0). Ensure you replace the `id` with the one obtained in the token creation (step 3).
- *Secret Token*: the OBS's token string obtained in step 3.
- *Trigger*. We just need you to select `Merge Requests`.

As soon as you add the webhook, OBS will start receiving requests for every Merge Request happening in GitLab.

> ### 5. The Workflow is Triggered in OBS

Reviewer 2: ```suggestion
### 5. The Workflow Is Triggered in OBS
```

An increasing number of people want to handle their package sources with those systems too
and while there is an integration between SCMs and OBS, it could definitely be improved.
In today's post, we are sharing how we began to move to meet that need.

## SCM to Collaborate, OBS to Build

We have started to integrate GitHub and GitLab for a scenario that seems to be the 
most expected one among OBS users: managing the package sources in an SCM system
and use OBS to build the source code and report back the results to GitHub or GitLab.

We propose the following steps to achieve this goal. Give them a try!

### 1. Set a Configuration File

In the root directory of your GitHub/GitLab repository, create a directory `.obs` which contains the configuration file `workflows.yml`.

The content of the file should look like this:

```
workflow1:
  filters:
    event: pull_request
    branches:
      only:
        - master
        - improvements
      ignore:
        - '2.10'
  steps:
    - branch_package:
        source_project: home:Iggy
        source_package: ctris
        destination_project_prefix: home:Iggy:PR- 
        destination_package: ctris
```

> For now, OBS only accepts one workflow with a single step: `branch_package`.

Reviewer 3: Nitpick: I would explain here that the destination project will be the concatenation of the value of `destination_project_prefix` and the number of the pull request as suffix.

Author: I rephrased the sentence to clarify that.

I also removed the `filter` part on the .yml I use like an example.

  <figcaption>GitHub Webhook Form</figcaption>
</figure>

You have to fill in the form with
- *Payload URL*: the OBS endpoint to trigger the workflow operation (https://build.opensuse.org/trigger/workflow?id=0). Ensure you replace the `id` with the one obtained in the token creation (step 4).
- *Content type*: please choose `application/json`.
- *Secret*: the OBS's token string obtained in step 4.
- *Enable SSL verification*.
- *Let me select individual events*. We just need you to select `Pull requests`.

As soon as you add the webhook, OBS will start receiving requests for every Pull Request happening in GitHub.

#### Set a GitLab Webhook

In GitLab, you can go to *Settings -> Webhooks*.

<figure>
  <img src="/images/posts/sprint_95_gitlab_webhook_form.png" alt="GitLab Webhook Form" />
  <figcaption>GitLab Webhook Form</figcaption>
</figure>

You have to fill in the form with
- *URL*: the OBS endpoint to trigger the workflow operation (https://build.opensuse.org/trigger/workflow?id=0). Ensure you replace the `id` with the one obtained in the token creation (step 4).
- *Secret Token*: the OBS's token string obtained in step 4.

> - *Trigger*. We just need you to select `Merge Requests`.

Reviewer: The sreenshot doesn't reflect this

Author: Screenshot updated, now only "Merge Requests" is selected.
